test/api/core/notifier/checkNewRoutes.test.js

```diff
@@ -0,0 +1,75 @@
+/**
+ * This test is made to ensure that all known elasticsearch routes that should fire a notification are plugged
+ * to the notifier core component.
+ *
+ * This is to ensure that new routes are firing notifications if they are meant to.
+ */
+var
+  should = require('should'),
+  winston = require('winston'),
+  _ = require('lodash'),
+  rewire = require('rewire'),
+  RequestObject = require.main.require('lib/api/core/models/requestObject'),
+  ResponseObject = require.main.require('lib/api/core/models/responseObject'),
+  Notifier = rewire('../../../../lib/api/core/notifier'),
+  params = require('rc')('kuzzle'),
+  Kuzzle = require.main.require('lib/api/Kuzzle');
+
+
+describe('Test: notifier.workerNotification', function () {
+  var
+    kuzzle,
+    blacklist = ['init', 'search', 'get', 'count', 'deleteCollection', 'import', 'putMapping', 'getMapping', 'listCollections', 'reset'],
+    requestObject = new RequestObject({
+      controller: 'write',
+      action: '',
+      requestId: 'foo',
+      collection: 'bar',
+      body: { foo: 'bar' }
+    }),
+    responseObject = new ResponseObject(requestObject, {});
+
+  before(function () {
+    kuzzle = new Kuzzle();
+
+    kuzzle.log = new (winston.Logger)({transports: [new (winston.transports.Console)({level: 'silent'})]});
+    return kuzzle.start(params, {dummy: true});
+  });
+
+  it('should map every notifiable action to a notification handler', function (done) {
+    var
+      knownActions = [],
+      performedActions = [],
+      mockupAction = function (responseObject) {
+        performedActions.push(responseObject.action);
+        return Promise.resolve({});
+      };
+
+    this.timeout(100);
+
+    Notifier.__with__({
+      notifyDocumentCreate: mockupAction,
+      notifyDocumentUpdate: mockupAction,
+      notifyDocumentDelete: mockupAction
+    })(function () {
+      for (var action in kuzzle.services.list.readEngine) {
+        if (kuzzle.services.list.readEngine.hasOwnProperty(action) && typeof kuzzle.services.list.readEngine[action] === 'function') {
+          knownActions.push(action);
```

Comment:
+1


*Jeff - Kuzzle Team*
jeff@kuzzle.io

------------------------------
kuzzle.io
@kuzzleio <https://twitter.com/kuzzleio>
Fork it on GitHub <https://github.com/kuzzleio>

2015-11-04 17:57 GMT+01:00 Sébastien Cottinet <notifications@github.com>:

> In test/api/core/notifier/checkNewRoutes.test.js
> <https://github.com/kuzzleio/kuzzle/pull/61#discussion_r43906915>:
>
> > +      performedActions = [],
> > +      mockupAction = function (responseObject) {
> > +        performedActions.push(responseObject.action);
> > +        return Promise.resolve({});
> > +      };
> > +
> > +    this.timeout(100);
> > +
> > +    Notifier.__with__({
> > +      notifyDocumentCreate: mockupAction,
> > +      notifyDocumentUpdate: mockupAction,
> > +      notifyDocumentDelete: mockupAction
> > +    })(function () {
> > +      for (var action in kuzzle.services.list.readEngine) {
> > +        if (kuzzle.services.list.readEngine.hasOwnProperty(action) && typeof kuzzle.services.list.readEngine[action] === 'function') {
> > +          knownActions.push(action);
>
> It's to make the (knownActions - performedActions) array match exactly the
> blacklist one (line 66).
>
> It allows testing the following cases with one should() test:
>
>    - if an ES function is not blacklisted (should fail)
>    - if an ES function throws a notification AND is blacklisted in the
>    test (should fail)
>    - if an ES function has a notification and is not blacklisted (should
>    succeed)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kuzzleio/kuzzle/pull/61/files#r43906915>.
>